### PR TITLE
fix compare diffs in case of array of json

### PR DESCRIPTION
**Pull request by AhmedLotfy02**
## Related Issue
  - Compare diffs (gojsondiff) fails in case of array of json

Closes: #1659

#### Describe the changes you've made
- Check if it's an array then iterate over each item and compare diffs (some work around).

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, local variables)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Please let us know if any test cases are added

Please describe the tests(if any). Provide instructions how its affecting the coverage.

#### Describe if there is any unusual behaviour of your code(Write `NA` if there isn't)
A clear and concise description of it.

## Checklist:

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] New and existing unit tests pass locally with my changes.

### In case of normal json:
json
### In case of array of json:
array

### CODE REVIEW

**Comment by AhmedLotfy02** [comment]
@shivamsouravjha While debugging the **array of json** I saw that noise map contains some parameters which aren't printed in the expected and actual body as they are classified as noise.

File: pkg/service/test/output.go [hunk 1]
```diff
@@ -200,14 +200,64 @@ func diffIndex(s1, s2 string) (int, bool) {
...
+					return "", err
+				}
+				// Perform comparison for each item
+				diffString, err := performComparison(itemJSON1, itemJSON2)
+				if err != nil {
+					return "", err
+				}
+				returnedString += diffString
+			}
+		}
```
Comment (shivamsouravjha) [inline pkg/service/test/output.go:238]: we internally used something similar thus iteration two arrays twice doesn't seem to be a good idea.

There exists alternatives which can ensure we don't have to compare manually, the issue I created was to look for such alternatives

Comment (AhmedLotfy02) [reply pkg/service/test/output.go:238]: Ok I will search for it.

Comment (AhmedLotfy02) [reply pkg/service/test/output.go:238]: @shivamsouravjha  We can use **go-test** package especially deep compare function 
https://pkg.go.dev/github.com/go-test/deep@v1.0.8
I made  a sample code to test it with two objects
`[
  {
    "firstName": "John",
    "lastName": "Doe",
    "address2": "14th Main Street, Royal Street",
    "city": "London, England",
    "telephone": "12345678",
    "id": 1,
    "pets": []
  },
  {
    "firstName": "Joe",
    "lastName": "Hopkins",
    "address": "24th Main Street, Something Street",
    "city": "Los Angeles, California",
    "telephone": "9090812030",
    "id": 2,
    "pets": []
  }
]
`
<hr>

`
[
  {
    "firstName2": "John",
    "lastName": "Doe",
    "address2": "14th Main Street, Royal Street",
    "city": "London, England",
    "telephone": "12345678",
    "id": 1,
    "pets": []
  },
  {
    "firstName": "Joe",
    "lastName": "Hopkins",
    "address": "24th Main Street, Something Street",
    "city": "Los Angeles, California",
    "telephone": "9090812030",
    "id": 2,
    "pets": []
  }
]
`
<hr>
I extracted the fields using regex and then append +, - before fields (it's a simulation about what will happen in keploy code).


`func TestDeepEqualJSONArrays() {`
`

	jsonArray1 := `[{"firstName":"John","lastName":"Doe","address2":"14th Main Street, Royal Street","city":"London, England","telephone":"12345678","id":1,"pets":[]},{"firstName":"Joe","lastName":"Hopkins","address":"24th Main Street, Something Street","city":"Los Angeles, California","telephone":"9090812030","id":2,"pets":[]}]`
	jsonArray2 := `[{"firstName2":"John","lastName":"Doe","address2":"14th Main Street, Royal Street","city":"London, England","telephone":"12345678","id":1,"pets":[]},{"firstName":"Joe","lastName":"Hopkins","address":"24th Main Street, Something Street","city":"Los Angeles, California","telephone":"9090812030","id":2,"pets":[]}]`

	var array1, array2 []map[string]interface{}

	// Unmarshal JSON arrays into slices of structs
	if err := json.Unmarshal([]byte(jsonArray1), &array1); err != nil {
		//print error
		fmt.Println("Error unmarshaling jsonArray1: %v", err)
	}
	if err := json.Unmarshal([]byte(jsonArray2), &array2); err != nil {
		fmt.Println("Error unmarshaling jsonArray1: %v", err)
	}


	// Compare arrays
	if diff := deep.Equal(array1, array2); diff != nil {
		re := regexp.MustCompile(`\[(\D\w*)\]:`)
		for _, d := range diff {

			if d[len(d)-1] == '>' {
				match := re.FindStringSubmatch(d)

				if len(match) > 1 {
					jsonArray1 = strings.Replace(jsonArray1, match[1], "+ "+match[1], 1)
				}
			} else {
				match := re.FindStringSubmatch(d)

				if len(match) > 1 {
					jsonArray2 = strings.Replace(jsonArray2, match[1], "- "+match[1], 1)
				}
			}
		}

		fmt.Println("Diff jsonArray1:", jsonArray1)
		fmt.Println("Diff jsonArray2:", jsonArray2)

	}
}
`

if it's good tell me to begin coding it in keploy

**Comment by gouravkrosx** [comment]
@AhmedLotfy02 Please update your branch according to the refactored changes.

**Pull request closed without merging**